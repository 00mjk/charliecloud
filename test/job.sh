#!/bin/bash

# This script evaluates the configuration and state of the virtual cluster. We
# test if everything is in order with the guest and host by comparing its
# output to reference output. To facilitate variable output, there is a
# comment facility; anything on a line matching '\s+//.*$' is removed before
# comparison.

exec 2>&1

export LC_ALL=C  # for consistent sorting, etc.
export PATH=$PATH:/sbin

function sec () {
    echo
    echo "$@"
}


sec '#### Test initialization ####'

sec '* distribution ID'
if [ -f /etc/debian_version ]; then
    echo 'ok  // Debian or derivative'
    DEBIAN=yes
elif [ -f /etc/redhat-release ]; then
    echo 'ok  // Red Hat or derivative'
    REDHAT=yes
else
    echo 'error: unknown distribution, aborting'
    exit 1
fi

sec '* Charliecloud environment variables used in this script'
if [ -d "$CH_DATA1" -a -n "$CH_GUEST_ID" ]; then
    mkdir -p $CH_DATA1/$CH_GUEST_ID
    OUTDIR=$CH_DATA1/$CH_GUEST_ID
    echo ok
else
    echo 'missing Charliecloud environment variables, aborting'
    exit 1
fi

sec '* needed binaries'
for i in ip lsb_release route; do
    echo -n "$i: "
    type -t $i || echo 'not found'
done

sec '* sudo without password'
if ( ! sudo -n echo 'ok' ); then
    echo 'sudo failed, aborting'
    exit 1
fi


sec '#### Installation ####'

sec '* root filesystem label'
sudo dumpe2fs /dev/vda1 2>&1 | fgrep 'Filesystem volume name' 2>&1

sec '* user charlie is present'
id -nu charlie


sec '#### Configuration ####'  # see also bootstrap.sh

sec '### Console and friends'

sec '* serial console'
tr '[:space:]' '\n' < /proc/cmdline | fgrep console= | sort

sec '* FANCYTTY off'
if [ "$DEBIAN" ]; then
    fgrep -q 'FANCYTTY=0' /etc/lsb-base-logging.sh && echo ok
elif [ "$REDHAT" ]; then
    echo 'ok  // not tested for Red Hat (issue #37)'
fi

sec '* framebuffer console disabled'
# http://kb.digium.com/articles/FAQ/How-to-disable-the-Linux-frame-buffer-if-it-s-causing-problems
dmesg | fgrep -i 'frame buffer'


sec '### Filesystem'

sec '* Environment variables'
set | fgrep -i proxy | sort > $OUTDIR/vars-proxy.actual
# trim final octets of MAC and IP addresses
set \
  | fgrep -i ch_ \
  | sort > $OUTDIR/vars-charlie.actual

sec '* /ch and subdirectories'
ls -1F /ch

sec '* /etc/fstab and filesystems mounted under /ch'
egrep '(^LABEL=)|(/ch)' /etc/fstab | sort
mount | fgrep /ch | sort


sec '### Network'

sec '* persistent network device names'
ls -1 /etc/udev/rules.d/*-persistent-net*.rules
cat /etc/udev/rules.d/75-persistent-net-generator.rules

sec '* interface configuration'
echo 'not tested: we test functionality instead'

sec '* symlinks to files in /ch/meta'
readlink -e /etc/hosts
readlink -e /etc/resolv.conf


sec '### Users and groups'

sec '* umask'
umask

#sec '* /sbin and /usr/sbin in $PATH'
#echo $PATH | tr : '\n' | egrep '^/sbin$'
#echo $PATH | tr : '\n' | egrep '^/usr/sbin$'

sec '* serial port permissions'
ls -l /dev/ttyS[012] | sed -r -e 's/[T-] /Z /' \
                              -e 's/[A-Za-z0-9 :]{12} \//[timestamp] \//'
id charlie | fgrep -q '(dialout)' && echo 'charlie is in dialout group'

sec '## Passwords & authentication'

sec '* SSH configuration'
for i in PubkeyAuthentication PasswordAuthentication PermitRootLogin; do
    egrep "^$i " /etc/ssh/sshd_config
done

sec '* /etc/sudoers'
# sudo itself tested above
sudo bash -c set | fgrep -i proxy | sort > $OUTDIR/vars-proxy-sudo.actual

sec '* charlie empty password'
sudo egrep '^charlie:' /etc/shadow | cut -d: -f1-2

sec '## SSH keys'

sec '* ~/.ssh contents'
ls -1 ~/.ssh/{authorized_keys,config,id_*}

sec '* authorized_keys'
cat ~/.ssh/authorized_keys \
  | fgrep charlie@ \
  | sed -r 's/(ssh-)(dsa|rsa) [A-Za-z0-9+/=]+ charlie@([a-z0-9]+)$/\1[xsa] [key] charlie@[host]/'

sec '* config'
egrep -v '^#' ~/.ssh/config


sec '### Miscellaneous'

sec '* MPI hostfile configuration'
egrep '^orte_default_hostfile' /etc/openmpi/openmpi-mca-params.conf


sec '#### Runtime tests ####'

sec '### Filesystems'
# fstab and mounts checked above

sec '## /ch directories'

sec '* /ch/meta'
# This is not comprehensive. It just checks if a few key files are present.
ls -1 /ch/meta/{guest-macs,hostfile,hosts,resolv.conf}

sec '* /ch/opt'
ls -1 /ch/opt/$(lsb_release --short --codename)

sec '* /ch/tmp is writeable'
echo 'hello world' > /ch/tmp/test
sync
ls -1 /ch/tmp

sec '* /ch/data[1234]'
echo 'no additional tests'

sec '* swap space present'
cat /proc/swaps


sec '### Networking'

mac=$(ip addr show eth0 | sed -rn s'/^.* link\/ether ([0-9a-z:]+).*$/\1/p')
ip=$(ip addr show eth0 | sed -rn s'/^.* inet ([0-9.]+).*$/\1/p')
mask=$(ip addr show eth0 | sed -rn s'/^.* inet [0-9.]+(\/[0-9.]+).*$/\1/p')
bcast=$(ip addr show eth0 | sed -rn s'/^.* brd ([0-9.]+).*$/\1/p')

sec '* hostname'
hostname > $OUTDIR/hostname.actual

sec '* MAC and IP space'
echo $mac | sed -r 's/(:[0-9a-z]+){2}$/:xx:xx/'
echo $ip | sed -r 's/(\.[0-9]{1,3}){2}$/.xxx.xxx/'
echo $mask
echo $bcast | sed -r 's/(\.[0-9]{1,3}){2}$/.xxx.xxx/'

sec '* MAC and IP match'
diff -u <(echo $mac) <(echo -n 0c:00; printf ':%02x' ${ip//./ }; echo)

sec '* MAC matches /ch/meta/guest-macs'
diff -ui <(echo $mac) <(sed -nr "s/^$CH_GUEST_ID (.+)$/\1/p" /ch/meta/guest-macs)

sec '* routing table'
ip route > $OUTDIR/route.actual

sec '* /etc/hosts link'
readlink /etc/hosts

sec '* /etc/resolv.conf link'
readlink /etc/resolv.conf

sec '* can we reach the internet?'
# Note that this is a little bit of an odd duck, because there is currently no
# way to tell whether the network *should* be reachable. Therefore, we're
# going to just print the results later and let the user interpret them.
#
# We try www.google.com by name and IP so that we know whether DNS is getting
# in the way.
WGET="wget -nv --tries=1 --no-cookies --timeout=5 -O /dev/null"
$WGET http://216.58.216.228 2> $OUTDIR/wget.net
$WGET http://www.google.com 2>> $OUTDIR/wget.net
unset WGET

sec '* ssh other nodes'
for i in $(cat /ch/meta/guests); do
    ssh $i hostname
done

sec '* simple MPI job'
mpirun hostname | sort

sec '* listening ports'
sudo netstat -tulp | sed -r 's/[0-9]+\//[pid]\//'

sec '### Users and groups'

sec '* are we charlie?'
whoami

sec '* charlie users and groups'
id charlie \
  | sed -r 's/groups=//' \
  | tr ' ,' '\n' \
  | egrep '^[a-z]|[0-9]{4,}|[5-9][0-9]{2}' \
  | sort \
  > $OUTDIR/id-charlie.actual

sec '### Miscellaneous'

sec '* running virtualized?'
dmesg | fgrep 'QEMU Standard PC'

sec '* job stderr'
echo 'hello stderr' > /dev/ttyS2

sec '* virtio drivers in use?'
lspci | fgrep -i virtio | sed -r 's/^[0-9a-f.:]+ /[pciaddr] /' | sort

sec '* running processes not whitelisted'
ps axo args= \
  | egrep -v '^(\[|-?bash|cut|e?grep|init|less|orted|ps|sshd:?|sort|su|tee|udevd)' \
  | egrep -v '^(\(sd-pam)\)' \
  | egrep -v '^(/s?bin/(agetty|bash|getty|init|login|sh))' \
  | egrep -v '^(/usr/s?bin/(acpid|atd|cron|dbus-daemon|irqbalance|r?syslogd|sshd))' \
  | egrep -v '^(/lib/systemd/systemd(-(journald|logind|udevd))?)' \
  | cut -d' ' -f1 \
  | sort -u
