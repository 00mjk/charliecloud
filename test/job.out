
#### Test initialization ####

* distribution ID
ok  // Debian or derivative

* Charliecloud environment variables used in this script
ok

* needed binaries
ip: file
lsb_release: file
route: file

* sudo without password
ok

#### Installation ####

* root filesystem label
Filesystem volume name:   root

* user charlie is present
charlie

#### Configuration ####

### Console and friends

* serial console
console=tty0
console=ttyS0

* FANCYTTY off
ok

* framebuffer console disabled

### Filesystem

* Environment variables

* /ch and subdirectories
data1/
data2/
data3/
data4/
meta/
opt/
tmp/

* /etc/fstab and filesystems mounted under /ch
LABEL=root    /              ext4         noatime,errors=remount-ro         0 1
LABEL=tmp     /ch/tmp        ext4         noauto,noatime,errors=remount-ro  0 2
data1  /ch/data1  9p  nofail,noatime,trans=virtio,version=9p2000.L  0 0
data2  /ch/data2  9p  nofail,noatime,trans=virtio,version=9p2000.L  0 0
data3  /ch/data3  9p  nofail,noatime,trans=virtio,version=9p2000.L  0 0
data4  /ch/data4  9p  nofail,noatime,trans=virtio,version=9p2000.L  0 0
meta   /ch/meta   9p  nofail,noatime,trans=virtio,version=9p2000.L  0 0
opt    /ch/opt    9p  nofail,noatime,trans=virtio,version=9p2000.L  0 0
/dev/vdb2 on /ch/tmp type ext4 (rw,noatime,errors=remount-ro)
data1 on /ch/data1 type 9p (rw,noatime,sync,dirsync,trans=virtio,version=9p2000.L)
data2 on /ch/data2 type 9p (rw,noatime,sync,dirsync,trans=virtio,version=9p2000.L)
data3 on /ch/data3 type 9p (rw,noatime,sync,dirsync,trans=virtio,version=9p2000.L)
data4 on /ch/data4 type 9p (rw,noatime,sync,dirsync,trans=virtio,version=9p2000.L)
meta on /ch/meta type 9p (rw,noatime,sync,dirsync,trans=virtio,version=9p2000.L)
opt on /ch/opt type 9p (rw,noatime,sync,dirsync,trans=virtio,version=9p2000.L)

### Network

* persistent network device names
/etc/udev/rules.d/75-persistent-net-generator.rules

* interface configuration
not tested: we test functionality instead

* symlinks to files in /ch/meta
/ch/meta/hosts
/ch/meta/resolv.conf

### Users and groups

* umask
0007

* serial port permissions
crw--w---Z 1 root tty     4, 64 [timestamp] /dev/ttyS0
crw-rw---Z 1 root dialout 4, 65 [timestamp] /dev/ttyS1
crw-rw---Z 1 root dialout 4, 66 [timestamp] /dev/ttyS2
charlie is in dialout group

## Passwords & authentication

* SSH configuration
PubkeyAuthentication yes
PasswordAuthentication no
PermitRootLogin no

* /etc/sudoers

* charlie empty password
charlie:

## SSH keys

* ~/.ssh contents
/home/charlie/.ssh/authorized_keys
/home/charlie/.ssh/config
/home/charlie/.ssh/id_rsa
/home/charlie/.ssh/id_rsa.pub

* authorized_keys
from="172.22.*" ssh-[xsa] [key] charlie@[host]

* config
Host 172.22.* chgu*
    UserKnownHostsFile /dev/null
    StrictHostKeyChecking no
    LogLevel error

### Miscellaneous

* MPI hostfile configuration
orte_default_hostfile = /ch/meta/hostfile

#### Runtime tests ####

### Filesystems

## /ch directories

* /ch/meta
/ch/meta/guest-macs
/ch/meta/hostfile
/ch/meta/hosts
/ch/meta/resolv.conf

* /ch/opt
10-hostname.sh
20-storage.sh
30-users.py
80-sync.sh
90-printvars.sh
99-runjob.sh
boot.sh
charlie.sh
network.sh
util.sh

* /ch/tmp is writeable
lost+found
test

* /ch/data[1234]
no additional tests

* swap space present
Filename				Type		Size	Used	Priority
/dev/vdb1                               partition	2097148	0	-1

### Networking

* hostname

* MAC and IP space
0c:00:ac:16:xx:xx
172.22.xxx.xxx
/24
172.22.xxx.xxx

* MAC and IP match

* MAC matches /ch/meta/guest-macs

* routing table

* /etc/hosts link
/ch/meta/hosts

* /etc/resolv.conf link
/ch/meta/resolv.conf

* can we reach the internet?

* ssh other nodes
chgu0
chgu1
chgu2

* simple MPI job
chgu0
chgu0
chgu1
chgu1
chgu2
chgu2

* listening ports
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 *:ssh                   *:*                     LISTEN      [pid]/sshd        
tcp6       0      0 [::]:ssh                [::]:*                  LISTEN      [pid]/sshd        

### Users and groups

* are we charlie?
charlie

* charlie users and groups

### Miscellaneous

* running virtualized?
[    0.000000] DMI: QEMU Standard PC

* job stderr

* virtio drivers in use?
[pciaddr] Ethernet controller: Red Hat, Inc Virtio network device
[pciaddr] SCSI storage controller: Red Hat, Inc Virtio block device
[pciaddr] SCSI storage controller: Red Hat, Inc Virtio block device
[pciaddr] Unclassified device [0002]: Red Hat, Inc Virtio filesystem
[pciaddr] Unclassified device [0002]: Red Hat, Inc Virtio filesystem
[pciaddr] Unclassified device [0002]: Red Hat, Inc Virtio filesystem
[pciaddr] Unclassified device [0002]: Red Hat, Inc Virtio filesystem
[pciaddr] Unclassified device [0002]: Red Hat, Inc Virtio filesystem
[pciaddr] Unclassified device [0002]: Red Hat, Inc Virtio filesystem

* running processes not whitelisted
